File: ios/RCTImageMarker/InvisibleWatermarkOptions.swift
import Foundation
import React

@available(iOS 13.0, *)
final class InvisibleWatermarkOptions: Options {
    let payload: String?
    let key: String
    let strength: String
    let search: String

    override init(dicOpts opts: [AnyHashable: Any]) throws {
        self.payload = opts["payload"] as? String
        self.key = opts["key"] as? String ?? ""
        self.strength = opts["strength"] as? String ?? "balanced"
        self.search = opts["search"] as? String ?? "fast"
        try super.init(dicOpts: opts)

        guard key.utf8.count >= InvisibleWatermark.minimumKeyBytes else {
            throw NSError(
                domain: ErrorDomainEnum.PARAMS_INVALID.rawValue,
                code: 0,
                userInfo: [NSLocalizedDescriptionKey: "key must contain at least \(InvisibleWatermark.minimumKeyBytes) UTF-8 bytes"]
            )
        }
        guard ["subtle", "balanced", "robust"].contains(strength) else {
            throw NSError(
                domain: ErrorDomainEnum.PARAMS_INVALID.rawValue,
                code: 0,
                userInfo: [NSLocalizedDescriptionKey: "Unsupported invisible watermark strength: \(strength)"]
            )
        }
        guard ["fast", "robust"].contains(search) else {
            throw NSError(
                domain: ErrorDomainEnum.PARAMS_INVALID.rawValue,
                code: 0,
                userInfo: [NSLocalizedDescriptionKey: "Unsupported invisible watermark search mode: \(search)"]
            )
        }
    }

    func requiredPayload() throws -> String {
        guard let payload else {
            throw NSError(
                domain: ErrorDomainEnum.PARAMS_REQUIRED.rawValue,
                code: 0,
                userInfo: [NSLocalizedDescriptionKey: "payload is required"]
            )
        }
        guard (1...InvisibleWatermark.maximumPayloadBytes).contains(payload.utf8.count) else {
            throw NSError(
                domain: ErrorDomainEnum.PARAMS_INVALID.rawValue,
                code: 0,
                userInfo: [NSLocalizedDescriptionKey: "payload must contain between 1 and \(InvisibleWatermark.maximumPayloadBytes) UTF-8 bytes"]
            )
        }
        return payload
    }

    static func checkEmbed(
        _ opts: [AnyHashable: Any],
        rejecter reject: @escaping RCTPromiseRejectBlock
    ) -> InvisibleWatermarkOptions? {
        check(opts, rejecter: reject, needsPayload: true)
    }

    static func checkDetect(
        _ opts: [AnyHashable: Any],
        rejecter reject: @escaping RCTPromiseRejectBlock
    ) -> InvisibleWatermarkOptions? {
        check(opts, rejecter: reject, needsPayload: false)
    }

    private static func check(
        _ opts: [AnyHashable: Any],
        rejecter reject: @escaping RCTPromiseRejectBlock,
        needsPayload: Bool
    ) -> InvisibleWatermarkOptions? {
        do {
            let options = try InvisibleWatermarkOptions(dicOpts: opts)
            if needsPayload { _ = try options.requiredPayload() }
            return options
        } catch let error as NSError {
            reject(error.domain, error.localizedDescription, error)
            return nil
        }
    }
}

File: android/src/main/java/com/jimmydaddy/imagemarker/base/InvisibleWatermarkOptions.kt
package com.jimmydaddy.imagemarker.base

import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReadableMap
import com.jimmydaddy.imagemarker.InvisibleWatermark

class InvisibleWatermarkOptions(options: ReadableMap) : Options(options) {
  val payload: String? = if (options.hasKey("payload")) options.getString("payload") else null
  val key: String = options.getString("key") ?: ""
  val strength: String = if (options.hasKey("strength")) options.getString("strength") ?: "balanced" else "balanced"
  val search: String = if (options.hasKey("search")) options.getString("search") ?: "fast" else "fast"

  init {
    val keyBytes = key.toByteArray(Charsets.UTF_8)
    if (keyBytes.size < InvisibleWatermark.MIN_KEY_BYTES) {
      throw MarkerError(ErrorCode.INVALID_PARAMS, "key must contain at least ${InvisibleWatermark.MIN_KEY_BYTES} UTF-8 bytes")
    }
    if (strength !in setOf("subtle", "balanced", "robust")) {
      throw MarkerError(ErrorCode.INVALID_PARAMS, "Unsupported invisible watermark strength: $strength")
    }
    if (search !in setOf("fast", "robust")) {
      throw MarkerError(ErrorCode.INVALID_PARAMS, "Unsupported invisible watermark search mode: $search")
    }
  }

  fun requirePayload(): String {
    val value = payload ?: throw MarkerError(ErrorCode.PARAMS_REQUIRED, "payload is required")
    val size = value.toByteArray(Charsets.UTF_8).size
    if (size !in 1..InvisibleWatermark.MAX_PAYLOAD_BYTES) {
      throw MarkerError(
        ErrorCode.INVALID_PARAMS,
        "payload must contain between 1 and ${InvisibleWatermark.MAX_PAYLOAD_BYTES} UTF-8 bytes"
      )
    }
    return value
  }

  companion object {
    fun checkEmbed(opts: ReadableMap, promise: Promise): InvisibleWatermarkOptions? =
      check(opts, promise, requirePayload = true)

    fun checkDetect(opts: ReadableMap, promise: Promise): InvisibleWatermarkOptions? =
      check(opts, promise, requirePayload = false)

    private fun check(
      opts: ReadableMap,
      promise: Promise,
      requirePayload: Boolean
    ): InvisibleWatermarkOptions? {
      return try {
        InvisibleWatermarkOptions(opts).also { if (requirePayload) it.requirePayload() }
      } catch (error: Exception) {
        val markerError = MarkerError.fromInvalidParams(error, "Invalid invisible watermark options")
        promise.reject(markerError.getErrorCode(), markerError.getErrMsg())
        null
      }
    }
  }
}
